presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java

```diff
@@ -1911,7 +1911,9 @@ public ConnectorPushdownFilterResult pushdownFilter(ConnectorSession session, Co
         domainPredicate.getDomains().get().keySet().stream()
                 .map(Subfield::getRootName)
                 .forEach(predicateColumnNames::add);
-        extractAll(decomposedFilter.getRemainingExpression()).stream()
+        // optimize the remaining expression before adding to predicateColumns
+        RowExpression optimizedRemainingExpression = rowExpressionService.getExpressionOptimizer().optimize(decomposedFilter.getRemainingExpression(), OPTIMIZED, session);
+        extractAll(optimizedRemainingExpression).stream()
```

Comment:
I'm seeing the following test case failing. It used to work because PickTableLayout extract tuple domain and checked for isNone. I don't think that's a proper way though. It seems like HiveMetadata.pushdownFilter needs to check decomposedFilter.getTupleDomain() for isNone.

```
        assertPlan(pushdownFilterEnabled(), "SELECT linenumber FROM lineitem WHERE linenumber > 10 AND linenumber < 0",
                output(values("linenumber")));
```

Reply:
yes looks like the optmization does not do a good job there, so I will put my previous version

Reply:
@bhhari The previous version is masking the bug in the connector, but it doesn't fix it. The connector must handle properly the case when tuple domain part of the expression matches nothing (e.g. isNone() == true). Currently, it treats it as if any value matches. You can see that the bug is still there, by changing the query to use range filters on struct subfields:

```
SELECT id FROM test WHERE x.a > 10 AND x.a < 0
```

Here, the connector simply drops the filter as if it is always true.

A proper fix could be to 

- [Hive connector] check isNone on tuple domain and `return new ConnectorPushdownFilterResult(getTableLayout(session, currentLayoutHandle.get()), FALSE_CONSTANT);`
- [Hive connector] check if optimized remaining expression is always false and  `return new ConnectorPushdownFilterResult(getTableLayout(session, currentLayoutHandle.get()), FALSE_CONSTANT);`
- [Engine/PickTableLayout] Optimize entire expression and, if always false, return ValuesNode without calling the connector. This optimization is optional though. There might not be any reason to put it in. E.g. the change might be to just remove the existing optimization that extracts tuple domain and checks it for isNone.